man/grok-dumb-pull.1
.\" Man page generated from reStructuredText.
.
.TH GROK-DUMB-PULL 1 "2020-08-14" "2.0.0" ""
.SH NAME
GROK-DUMB-PULL \- Update git repositories not managed by grokmirror
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
grok\-dumb\-pull [options] /path/to/repos
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
This is a satellite utility that updates repositories not exported via
grokmirror manifest. You will need to manually clone these repositories
using "git clone \-\-mirror" and then define a cronjob to update them as
frequently as you require. Grok\-dumb\-pull will bluntly execute "git
remote update" in each of them.
.SH OPTIONS
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.TP
.B \-\-version
show program\(aqs version number and exit
.TP
.B \-h\fP,\fB  \-\-help
show this help message and exit
.TP
.B \-v\fP,\fB  \-\-verbose
Be verbose and tell us what you are doing
.TP
.B \-s\fP,\fB  \-\-svn
The remotes for these repositories are Subversion
.TP
.BI \-r \ REMOTES\fP,\fB \ \-\-remote\-names\fB= REMOTES
Only fetch remotes matching this name (accepts globbing,
can be passed multiple times)
.TP
.BI \-u \ POSTHOOK\fP,\fB \ \-\-post\-update\-hook\fB= POSTHOOK
Run this hook after each repository is updated. Passes
full path to the repository as the sole argument.
.TP
.BI \-l \ LOGFILE\fP,\fB \ \-\-logfile\fB= LOGFILE
Put debug logs into this file
.UNINDENT
.UNINDENT
.UNINDENT
.SH EXAMPLES
.sp
The following will update all bare git repositories found in
/path/to/repos hourly, and /path/to/special/repo.git daily, fetching
only the "github" remote:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
MAILTO=root
# Update all repositories found in /path/to/repos hourly
0 * * * * mirror /usr/bin/grok\-dumb\-pull /path/to/repos
# Update /path/to/special/repo.git daily, fetching "github" remote
0 0 * * * mirror /usr/bin/grok\-dumb\-pull \-r github /path/to/special/repo.git
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Make sure the user "mirror" (or whichever user you specified) is able to
write to the repos specified.
.SH SEE ALSO
.INDENT 0.0
.IP \(bu 2
grok\-pull(1)
.IP \(bu 2
grok\-manifest(1)
.IP \(bu 2
grok\-fsck(1)
.IP \(bu 2
git(1)
.UNINDENT
.SH SUPPORT
.sp
Email \fI\%tools@linux.kernel.org\fP\&.
.SH AUTHOR
mricon@kernel.org

License: GPLv3+
.SH COPYRIGHT
The Linux Foundation and contributors
.\" Generated by docutils manpage writer.
.


CHANGELOG.rst
v2.0.0 (2020-08-14)
-------------------
- Major rewrite to improve shared object storage. See UPGRADING.rst for
  upgrade strategy. Below are major highlights.
- Drop support for python < 3.6
- Introduce "object storage" repositories that benefit from git-pack
  delta islands
- Make grok-pull operate in daemon mode (with -o) (see contrib for
  systemd unit files)
- Provide a socket listener for pubsub push updates (see contrib for
  Google pubsubv1.py)
- Merge fsck.conf and repos.conf into a single config file
- Grok-manifest will now record where HEAD is pointing so replicas can
  properly track these changes.


v1.2.2 (2019-10-23)
-------------------
- Small bugfixes
- Generate commit-graph file if the version of git is new
  enough to support it. This is done during grok-fsck any time we
  decide that the repository needs to be repacked. You can force
  this off by setting commitgraph=never in config.


v1.2.1 (2019-03-11)
-------------------
- Minor feature improvement changing how precious=yes works.
  Grokmirror will now turn preciousObjects off for the duration
  of the repack. We still protect shared repositories against
  inadvertent object pruning by outside processes, but this
  allows us to clean up loose objects and obsolete packs.
  To have the 1.2.0 behaviour back, set precious=always, but it
  is only really useful in very rare cases.


v1.2.0 (2019-02-14)
-------------------
- Make sure to set gc.auto=0 on repositories to avoid pruning repos
  that are acting as alternates to others. We run our own prune
  during fsck, so there is no need to auto-gc, ever (unless you
  didn't set up grok-fsck, in which case you're not doing it right).
- Rework the repack code to be more clever -- instead of repacking
  based purely on dates, we now track the number of loose objects
  and the number of generated packs. Many of the settings are
  hardcoded for the moment while testing, but will probably end up
  settable via global and per-repository config settings.
- The following fsck.conf settings have no further effect:
    - repack_flags (replaced with extra_repack_flags)
    - full_repack_flags (replaced with extra_repack_flags_full)
    - full_repack_every (we now figure it out ourselves)
- Move git command invocation routines into a central function to
  reduce the amount of code duplication. You can also set the path
  to the git binary using the GITBIN env variable or by simply
  adding it to your path.
- Add "reclone_on_errors" setting in fsck.conf. If fsck/repack/prune
  comes across a matching error, it will mark the repository for
  recloning and it will be cloned anew from the master the next time
  grok-pull runs. This is useful for auto-correcting corruption on the
  mirrors. You can also manually request a reclone by creating a
  "grokmirror.reclone" file in a repository.
- Set extensions.preciousObjects for repositories used with git
  alternates if precious=yes is set in fsck.conf. This helps further
  protect shared repos from erroneous pruning (e.g. done manually by
  an administrator).


v1.1.1 (2018-07-25)
-------------------
- Quickfix a bug that was causing repositories to never be repacked
  due to miscalculated fingerprints.


v1.1.0 (2018-04-24)
-------------------
- Make Python3 compatible (thanks to QuLogic for most of the work)
- Rework grok-fsck to improve functionality:

  - run repack and prune before fsck, for optimal safety
  - add --connectivity flag to run fsck with --connectivity-only
  - add --repack-all-quick to trigger a quick repack of all repos
  - add --repack-all-full to trigger a full repack of all repositories
    using the defined full_repack_flags from fsck.conf
  - always run fsck with --no-dangling, because mirror admins are not
    responsible for cleaning those up anyway
  - no longer locking repos when running repack/prune/fsck, because
    these operations are safe as long as they are done by git itself

- fix grok-pull so it no longer purges repos that are providing
  alternates to others
- fix grok-fsck so it's more paranoid when pruning repos providing
  alternates to others (checks all repos on disk, not just manifest)
- in verbose mode, most commands will draw progress bars (handy with
  very large connections of repositories)


grok-dumb-pull.py
#!/usr/bin/python -tt
# Copyright (C) 2013 by The Linux Foundation and contributors
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import os
import sys

import grokmirror
import logging
import time
import fnmatch
import subprocess
import shutil
import calendar

from fcntl import lockf, LOCK_EX, LOCK_UN, LOCK_NB

from git import Repo

logger = logging.getLogger(__name__)

def git_rev_parse_all(gitdir):
    logger.debug('Running: GIT_DIR=%s git rev-parse --all', gitdir)

    env  = {'GIT_DIR': gitdir}
    args = ['git', 'rev-parse', '--all']

    (output, error) = subprocess.Popen(args, stdout=subprocess.PIPE,
        stderr=subprocess.PIPE, env=env).communicate()

    error = error.strip()

    if error:
        # Put things we recognize into debug
        debug = []
        warn  = []
        for line in error.split('\n'):
                warn.append(line)
        if debug:
            logger.debug('Stderr: %s' % '\n'.join(debug))
        if warn:
            logger.warning('Stderr: %s' % '\n'.join(warn))

    return output


def git_remote_update(args, env):
    logger.debug('Running: GIT_DIR=%s %s' % (env['GIT_DIR'], ' '.join(args)))

    (output, error) = subprocess.Popen(args, stdout=subprocess.PIPE,
        stderr=subprocess.PIPE, env=env).communicate()

    error = error.strip()

    if error:
        # Put things we recognize into debug
        debug = []
        warn  = []
        for line in error.split('\n'):
            if line.find('From ') == 0:
                debug.append(line)
            elif line.find('-> ') > 0:
                debug.append(line)
            else:
                warn.append(line)
        if debug:
            logger.debug('Stderr: %s' % '\n'.join(debug))
        if warn:
            logger.warning('Stderr: %s' % '\n'.join(warn))

def dumb_pull_repo(gitdir, remotes, svn=False):
    # verify it's a git repo and fetch all remotes
    logger.debug('Will pull %s with following remotes: %s' % (gitdir, remotes))
    try:
        repo = Repo(gitdir)
        assert repo.bare == True
    except:
        logger.critical('Error opening %s.' % gitdir)
        logger.critical('Make sure it is a bare git repository.')
        sys.exit(1)

    try:
        grokmirror.lock_repo(gitdir, nonblocking=True)
    except IOError, ex:
        logger.info('Could not obtain exclusive lock on %s' % gitdir)
        logger.info('\tAssuming another process is running.')
        return False

    env = {'GIT_DIR': gitdir}

    old_revs = git_rev_parse_all(gitdir)

    if svn:
        logger.debug('Using git-svn for %s' % gitdir)

        for remote in remotes:
            # arghie-argh-argh
            if remote == '*':
                remote = '--all'

            logger.info('Running git-svn fetch %s in %s' % (remote, gitdir))
            args = ['/usr/bin/git', 'svn', 'fetch', remote]
            git_remote_update(args, env)

    else:
        # Not an svn remote
        hasremotes = repo.git.remote()
        if not len(hasremotes.strip()):
            logger.info('Repository %s has no defined remotes!' % gitdir)
            return False

        logger.debug('existing remotes: %s' % hasremotes)
        for remote in remotes:
            remotefound = False
            for hasremote in hasremotes.split('\n'):
                if fnmatch.fnmatch(hasremote, remote):
                    remotefound = True
                    logger.debug('existing remote %s matches %s' % (
                        hasremote, remote))
                    args = ['/usr/bin/git', 'remote', 'update', hasremote]
                    logger.info('Updating remote %s in %s' % 
                                (hasremote, gitdir))

                    git_remote_update(args, env)
                    didwork = True

            if not remotefound:
                logger.info('Could not find any remotes matching %s in %s' % (
                    remote, gitdir))

    new_revs = git_rev_parse_all(gitdir)
    grokmirror.unlock_repo(gitdir)

    if old_revs == new_revs:
        logger.debug('No new revs, no updates')
        return False

    logger.debug('New revs found -- new content pulled')
    return True

def run_post_update_hook(hookscript, gitdir):
    if hookscript == '':
        return
    if not os.access(hookscript, os.X_OK):
        logger.warning('post_update_hook %s is not executable' % hookscript)
        return

    args = [hookscript, gitdir]
    logger.debug('Running: %s' % ' '.join(args))
    (output, error) = subprocess.Popen(args, stdout=subprocess.PIPE,
        stderr=subprocess.PIPE).communicate()

    error  = error.strip()
    output = output.strip()
    if error:
        # Put hook stderror into warning
        logger.warning('Hook Stderr: %s' % error)
    if output:
        # Put hook stdout into info
        logger.info('Hook Stdout: %s' % output)

if __name__ == '__main__':
    from optparse import OptionParser

    usage = '''usage: %prog [options] /path/to/repos
    Bluntly fetch remotes in specified git repositories.
    '''

    parser = OptionParser(usage=usage, version=grokmirror.VERSION)
    parser.add_option('-v', '--verbose', dest='verbose', action='store_true',
        default=False,
        help='Be verbose and tell us what you are doing')
    parser.add_option('-s', '--svn', dest='svn', action='store_true',
        default=False,
        help='The remotes for these repositories are Subversion')
    parser.add_option('-r', '--remote-names', dest='remotes', action='append',
        default=[],
        help='Only fetch remotes matching this name (accepts globbing, '
             'can be passed multiple times)')
    parser.add_option('-u', '--post-update-hook', dest='posthook',
        default='',
        help='Run this hook after each repository is updated. Passes '
             'full path to the repository as the sole argument.')
    parser.add_option('-l', '--logfile', dest='logfile',
        default=None,
        help='Put debug logs into this file')

    (opts, args) = parser.parse_args()

    if not len(args):
        parser.error('You must provide at least a path to the repos to pull')

    if not len(opts.remotes):
        opts.remotes = ['*']

    logger.setLevel(logging.DEBUG)

    ch = logging.StreamHandler()
    formatter = logging.Formatter('%(message)s')
    ch.setFormatter(formatter)

    if opts.verbose:
        ch.setLevel(logging.INFO)
    else:
        ch.setLevel(logging.CRITICAL)

    logger.addHandler(ch)

    if opts.logfile is not None:
        ch = logging.FileHandler(opts.logfile)
        formatter = logging.Formatter("[%(process)d] %(asctime)s - %(levelname)s - %(message)s")
        ch.setFormatter(formatter)

        ch.setLevel(logging.DEBUG)
        logger.addHandler(ch)

    # push our logger into grokmirror to override the default
    grokmirror.logger = logger

    # Find all repositories we are to pull
    for entry in args:
        if entry[-4:] == '.git':
            if not os.path.exists(entry):
                logger.critical('%s does not exist' % entry)
                continue

            logger.debug('Found %s' % entry)
            didwork = dumb_pull_repo(entry, opts.remotes, svn=opts.svn)
            if didwork:
                run_post_update_hook(opts.posthook, entry)

        else:
            logger.debug('Finding all git repos in %s' % entry)
            for founddir in grokmirror.find_all_gitdirs(entry):
                didwork = dumb_pull_repo(founddir, opts.remotes, svn=opts.svn)
                if didwork:
                    run_post_update_hook(opts.posthook, founddir)

